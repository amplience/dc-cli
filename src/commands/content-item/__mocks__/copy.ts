import { CopyItemBuilderOptions } from '../../../interfaces/copy-item-builder-options.interface';
import { ConfigurationParameters } from '../../configure';
import { Arguments } from 'yargs';

let outputIds: string[];
let forceFail = false;
export const calls: Arguments<CopyItemBuilderOptions & ConfigurationParameters>[] = [];

export const setOutputIds = (ids: string[]): void => {
  outputIds = ids;
};

export const setForceFail = (fail: boolean): void => {
  forceFail = fail;
};

export const handler = async (argv: Arguments<CopyItemBuilderOptions & ConfigurationParameters>): Promise<boolean> => {
  calls.push(argv);
  const idOut = argv.exportedIds as string[];
  idOut.push(...outputIds);

  return !forceFail;
};
