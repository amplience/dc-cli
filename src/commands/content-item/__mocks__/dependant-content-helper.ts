import { ContentDependancy } from '../../../common/content-item/content-dependancy-tree';

function dependancy(id: string): ContentDependancy {
  return {
    _meta: {
      schema: 'http://bigcontent.io/cms/schema/v1/core#/definitions/content-link',
      name: 'content-link'
    },
    contentType: 'https://dev-solutions.s3.amazonaws.com/DynamicContentTypes/Accelerators/blog.json',
    id: id
  };
}

function dependProps(itemProps: [string, string][]): object {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const result: any = {};
  itemProps.forEach(element => {
    result[element[0]] = dependancy(element[1]);
  });
  return result;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function dependsOn(itemIds: string[], itemProps?: [string, string][]): any {
  itemProps = itemProps || [];
  return {
    links: itemIds.map(id => dependancy(id)),
    ...dependProps(itemProps)
  };
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function dependantType(items: number): any {
  return {
    $schema: 'http://json-schema.org/draft-07/schema#',
    $id: 'http://superbasic.com',

    title: 'Title',
    description: 'Description',

    allOf: [
      {
        $ref: 'http://bigcontent.io/cms/schema/v1/core#/definitions/content'
      }
    ],

    required: ['valid'],
    type: 'object',
    properties: {
      links: {
        title: 'title',
        type: 'array',
        minItems: items,
        maxItems: items,
        items: {
          allOf: [
            { $ref: 'http://bigcontent.io/cms/schema/v1/core#/definitions/content-link' },
            {
              properties: {
                contentType: {
                  enum: ['*']
                }
              }
            }
          ]
        }
      }
    },
    propertyOrder: []
  };
}
