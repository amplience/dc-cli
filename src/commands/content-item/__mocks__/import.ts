import { ImportItemBuilderOptions } from '../../../interfaces/import-item-builder-options.interface';
import { ConfigurationParameters } from '../../configure';
import { Arguments } from 'yargs';

type ReturnType = boolean | 'throw';
let expectedReturn: ReturnType = true;

export const calls: Arguments<ImportItemBuilderOptions & ConfigurationParameters>[] = [];
export const setExpectedReturn = (value: ReturnType): void => {
  expectedReturn = value;
};

export const handler = async (
  argv: Arguments<ImportItemBuilderOptions & ConfigurationParameters>
): Promise<boolean> => {
  calls.push(argv);

  if (expectedReturn == 'throw') {
    throw new Error('Forced throw in test.');
  }

  return expectedReturn;
};
