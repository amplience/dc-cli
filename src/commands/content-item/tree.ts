import { getDefaultLogPath } from '../../common/log-helpers';
import { Argv, Arguments } from 'yargs';
import { join, extname, resolve } from 'path';
import { ConfigurationParameters } from '../configure';
import { lstat, readdir, readFile } from 'fs';
import { promisify } from 'util';

import { ContentItem, ContentRepository } from 'dc-management-sdk-js';
import {
  ContentDependancyTree,
  ItemContentDependancies,
  RepositoryContentItem
} from '../../common/content-item/content-dependancy-tree';
import { ContentMapping } from '../../common/content-item/content-mapping';

export const command = 'tree <dir>';

export const desc = 'Print a content dependency tree from content in the given folder.';

export const LOG_FILENAME = (platform: string = process.platform): string =>
  getDefaultLogPath('item', 'tree', platform);

export const builder = (yargs: Argv): void => {
  yargs.positional('dir', {
    type: 'string',
    describe: 'Path to the content items to build a tree from. Should be in the same format as an export.'
  });
};

interface TreeOptions {
  dir: string;
}

export const traverseRecursive = async (path: string, action: (path: string) => Promise<void>): Promise<void> => {
  const dir = await promisify(readdir)(path);

  dir.sort();

  for (let i = 0; i < dir.length; i++) {
    let contained = dir[i];
    contained = join(path, contained);
    const stat = await promisify(lstat)(contained);
    if (stat.isDirectory()) {
      await traverseRecursive(contained, action);
    } else {
      await action(contained);
    }
  }
};

export const prepareContentForTree = async (repo: {
  basePath: string;
  repo: ContentRepository;
}): Promise<ContentDependancyTree> => {
  const contentItems: RepositoryContentItem[] = [];
  const schemaNames = new Set<string>();

  await traverseRecursive(resolve(repo.basePath), async path => {
    // Is this valid content? Must have extension .json to be considered, for a start.
    if (extname(path) !== '.json') {
      return;
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let contentJSON: any;
    try {
      const contentText = await promisify(readFile)(path, { encoding: 'utf8' });
      contentJSON = JSON.parse(contentText);
    } catch (e) {
      console.error(`Couldn't read content item at '${path}': ${e.toString()}`);
      return;
    }

    schemaNames.add(contentJSON.body._meta.schema);

    contentItems.push({ repo: repo.repo, content: new ContentItem(contentJSON) });
  });

  return new ContentDependancyTree(contentItems, new ContentMapping());
};

type LineIndexFrom = number;
type LineIndexTo = number;
type CircularLink = [LineIndexFrom, LineIndexTo];
interface ParentReference {
  item: ItemContentDependancies;
  line: number;
}

export const firstSecondThird = (index: number, total: number): number => {
  return index == total - 1 ? 2 : index === 0 ? 0 : 1;
};

const fstPipes = ['├', '├', '└'];
const circularPipes = ['╗', '║', '╝'];
const circularLine = '═';

export class TreeBuilder {
  lines: string[] = [];
  circularLinks: CircularLink[] = [];

  constructor(public evaluated: Set<ItemContentDependancies>) {}

  addDependency(item: ItemContentDependancies, evalThis: ParentReference[], fst: number, prefix: string): boolean {
    const depth = evalThis.length - 1;
    const pipe = depth < 0 ? '' : fstPipes[fst] + '─ ';

    const circularMatch = evalThis.find(parent => parent.item == item);
    if (circularMatch) {
      this.lines.push(`${prefix}${pipe}*** (${item.owner.content.label})`);
      this.circularLinks.push([circularMatch.line, this.lines.length - 1]);
      return false;
    } else if (this.evaluated.has(item)) {
      if (depth > -1) {
        this.lines.push(`${prefix}${pipe}(${item.owner.content.label})`);
      }
      return false;
    } else {
      this.lines.push(`${prefix}${pipe}${item.owner.content.label}`);
    }

    evalThis.push({ item, line: this.lines.length - 1 });
    this.evaluated.add(item);

    const filteredItems = item.dependancies.filter(dep => dep.resolved);
    filteredItems.forEach((dep, index) => {
      const subFst = firstSecondThird(index, filteredItems.length);
      const subPrefix = depth == -1 ? '' : fst === 2 ? '   ' : '│  ';
      this.addDependency(dep.resolved as ItemContentDependancies, [...evalThis], subFst, prefix + subPrefix);
    });
    return true;
  }
}

export const fillWhitespace = (original: string, current: string, char: string, targetLength: number): string => {
  let position = original.length;
  let repeats = targetLength - original.length;

  // Replace existing whitespace characters
  while (position < current.length && repeats > 0) {
    if (current[position] != char && current[position] == ' ') {
      current = current.slice(0, position) + char + current.slice(position + 1);
    }

    position++;
    repeats--;
  }

  if (repeats > 0) {
    current += char.repeat(repeats);
  }

  return current;
};

export const printTree = (item: ItemContentDependancies, evaluated: Set<ItemContentDependancies>): boolean => {
  const builder = new TreeBuilder(evaluated);

  const result = builder.addDependency(item, [], 0, '');

  if (!result) return false;

  const circularLinks = builder.circularLinks;
  const lines = builder.lines.map(line => line + ' ');
  const modifiedLines = [...lines];

  // Render circular references.
  // These are drawn as pipes on the right hand side, from a start line to an end line.

  const maxWidth = Math.max(...lines.map(x => x.length));

  for (let i = 0; i < circularLinks.length; i++) {
    const link = circularLinks[i];
    let linkDist = maxWidth + 2;

    // Find overlapping circular links. Push the link out further if a previously drawn line is there.
    for (let j = 0; j < i; j++) {
      const link2 = circularLinks[j];
      if (link[0] <= link2[1] && link[1] >= link2[0]) {
        linkDist += 2;
      }
    }

    // Write the circular dependency lines into the tree.

    for (let ln = link[0]; ln <= link[1]; ln++) {
      const end = ln == link[0] || ln == link[1];
      const original = lines[ln];
      let current = modifiedLines[ln];

      current = fillWhitespace(original, current, end ? circularLine : ' ', linkDist);
      current += circularPipes[firstSecondThird(ln - link[0], link[1] - link[0] + 1)];

      modifiedLines[ln] = current;
    }
  }

  modifiedLines.forEach(line => console.log(line));
  console.log('');
  return true;
};

export const handler = async (argv: Arguments<TreeOptions & ConfigurationParameters>): Promise<void> => {
  const dir = argv.dir;

  const tree = await prepareContentForTree({ basePath: dir, repo: new ContentRepository() });

  // Print the items in the tree.
  // Keep a set of all items that have already been printed.
  // Starting at the highest level, print all dependencies on the tree.

  const evaluated = new Set<ItemContentDependancies>();

  for (let i = tree.levels.length - 1; i >= 0; i--) {
    const level = tree.levels[i];
    console.log(`=== LEVEL ${i + 1} (${level.items.length}) ===`);

    level.items.forEach(item => {
      printTree(item, evaluated);
    });
  }

  let topLevelPrints = 0;

  if (tree.circularLinks.length > 0) {
    console.log(`=== CIRCULAR (${tree.circularLinks.length}) ===`);

    tree.circularLinks.forEach(item => {
      if (printTree(item, evaluated)) {
        topLevelPrints++;
      }
    });
  }

  console.log(`Finished. Circular Dependencies printed: ${topLevelPrints}`);
};
