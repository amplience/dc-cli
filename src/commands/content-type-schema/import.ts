import { Arguments, Argv } from 'yargs';
import { ConfigurationParameters } from '../configure';
import { ContentTypeSchema, DynamicContent, Hub, ValidationLevel } from 'dc-management-sdk-js';
import dynamicContentClientFactory from '../../services/dynamic-content-client-factory';
import paginator from '../../common/dc-management-sdk-js/paginator';
import { createStream } from 'table';
import { streamTableOptions } from '../../common/table/table.consts';
import { TableStream } from '../../interfaces/table.interface';
import { ImportBuilderOptions } from '../../interfaces/import-builder-options.interface';
import chalk from 'chalk';
import { createContentTypeSchema } from './create.service';
import { updateContentTypeSchema } from './update.service';
import { ImportResult, loadJsonFromDirectory, UpdateStatus } from '../../services/import.service';
import { jsonResolver } from '../../common/import/json-resolver';

export const command = 'import <dir>';

export const desc = 'Import Content Type Schemas';

export interface SchemaOptions {
  validation: ValidationLevel;
}

export const builder = (yargs: Argv): void => {
  yargs.positional('dir', {
    describe: 'Directory containing Content Type Schema definitions',
    type: 'string'
  });
};

export const storedSchemaMapper = (
  schema: ContentTypeSchema,
  storedSchemas: ContentTypeSchema[]
): ContentTypeSchema => {
  const found = storedSchemas.find(stored => stored.schemaId === schema.schemaId);
  const mutatedSchema = found ? { ...schema, id: found.id } : schema;

  return new ContentTypeSchema(mutatedSchema);
};

export const doCreate = async (hub: Hub, schema: ContentTypeSchema): Promise<ContentTypeSchema> => {
  try {
    const createdSchemaType = await createContentTypeSchema(
      schema.body || '',
      schema.validationLevel || ValidationLevel.CONTENT_TYPE,
      hub
    );

    return createdSchemaType;
  } catch (err) {
    throw new Error(`Error registering content type schema with body: ${schema.body}\n\n${err}`);
  }
};

const equals = (a: ContentTypeSchema, b: ContentTypeSchema): boolean =>
  a.id === b.id && a.schemaId === b.schemaId && a.body === b.body;

export const doUpdate = async (
  client: DynamicContent,
  schema: ContentTypeSchema
): Promise<{ contentTypeSchema: ContentTypeSchema; updateStatus: UpdateStatus }> => {
  try {
    const retrievedSchema = await client.contentTypeSchemas.get(schema.id || '');
    if (equals(retrievedSchema, schema)) {
      return { contentTypeSchema: retrievedSchema, updateStatus: UpdateStatus.SKIPPED };
    }
    const updatedSchema = await updateContentTypeSchema(
      retrievedSchema,
      schema.body || '',
      schema.validationLevel || ValidationLevel.CONTENT_TYPE
    );

    return { contentTypeSchema: updatedSchema, updateStatus: UpdateStatus.UPDATED };
  } catch (err) {
    throw new Error(`Error updating content type schema ${schema.schemaId || '<unknown>'}: ${err.message}`);
  }
};

export const processSchemas = async (
  schemasToProcess: ContentTypeSchema[],
  client: DynamicContent,
  hub: Hub
): Promise<void> => {
  const tableStream = (createStream(streamTableOptions) as unknown) as TableStream;

  tableStream.write([chalk.bold('ID'), chalk.bold('Schema ID'), chalk.bold('Result')]);
  for (const schema of schemasToProcess) {
    let schemaId = schema.id;
    let status: ImportResult;
    if (schema.id) {
      const result = await doUpdate(client, schema);
      status = result.updateStatus === UpdateStatus.SKIPPED ? 'UP-TO-DATE' : 'UPDATED';
    } else {
      const result = await doCreate(hub, schema);
      schemaId = result.id || 'UNKNOWN';
      status = 'CREATED';
    }
    tableStream.write([schemaId || '', schema.schemaId || '', status]);
  }
  process.stdout.write('\n');
};

export const resolveSchemaBody = async (schemas: ContentTypeSchema[], dir: string): Promise<ContentTypeSchema[]> => {
  for (const schema of schemas) {
    if (schema.body) {
      schema.body = await jsonResolver(schema.body, dir);
    }
  }
  return schemas;
};

export const handler = async (argv: Arguments<ImportBuilderOptions & ConfigurationParameters>): Promise<void> => {
  const { dir } = argv;
  const client = dynamicContentClientFactory(argv);
  const hub = await client.hubs.get(argv.hubId);
  const schemas = Object.values(loadJsonFromDirectory<ContentTypeSchema>(dir, ContentTypeSchema));
  await resolveSchemaBody(schemas, dir);
  const storedSchemas = await paginator(hub.related.contentTypeSchema.list);
  const schemasToProcess = schemas.map(schemas => storedSchemaMapper(schemas, storedSchemas));

  await processSchemas(schemasToProcess, client, hub);
};
