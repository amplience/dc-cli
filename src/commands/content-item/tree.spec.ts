// Copy tests are rather simple since they most of the work is done by import/export.
// Unique features are revert, throwing when parameters are wrong/missing,
// and forwarding input parameters to both import and export.

import { builder, command, handler, firstSecondThird, fillWhitespace, LOG_FILENAME } from './tree';
import Yargs from 'yargs/yargs';

import { writeFile } from 'fs';
import { join } from 'path';
import { promisify } from 'util';

import { ensureDirectoryExists } from '../../common/import/directory-utils';
import rmdir from 'rimraf';
import { getDefaultLogPath } from '../../common/log-helpers';

import { ItemTemplate } from '../../common/dc-management-sdk-js/mock-content';
import { dependsOn } from '../../commands/content-item/__mocks__/dependant-content-helper';
import { ContentItem, Status } from 'dc-management-sdk-js';

jest.mock('../../services/dynamic-content-client-factory');
jest.mock('../../common/log-helpers');

const consoleLogSpy = jest.spyOn(console, 'log');
const consoleErrorSpy = jest.spyOn(console, 'error');

function rimraf(dir: string): Promise<Error> {
  return new Promise((resolve): void => {
    rmdir(dir, resolve);
  });
}

describe('content-item tree command', () => {
  afterEach((): void => {
    jest.resetAllMocks();
  });

  it('should command should defined', function() {
    expect(command).toEqual('tree <dir>');
  });

  describe('builder tests', function() {
    it('should configure yargs', function() {
      const argv = Yargs(process.argv.slice(2));
      const spyPositional = jest.spyOn(argv, 'positional').mockReturnThis();

      builder(argv);

      expect(spyPositional).toHaveBeenCalledWith('dir', {
        type: 'string',
        describe: 'Path to the content items to build a tree from. Should be in the same format as an export.'
      });
    });
  });

  describe('firstSecondThird tests', function() {
    it('should return 0 for the first item in a list, above size 1', () => {
      expect(firstSecondThird(0, 2)).toEqual(0);
      expect(firstSecondThird(0, 3)).toEqual(0);
      expect(firstSecondThird(0, 4)).toEqual(0);
    });

    it('should return 2 for the last item in a list', () => {
      expect(firstSecondThird(0, 1)).toEqual(2);
      expect(firstSecondThird(1, 2)).toEqual(2);
      expect(firstSecondThird(2, 3)).toEqual(2);
      expect(firstSecondThird(3, 4)).toEqual(2);
    });

    it('should return 1 for any middle item in a list, above size 2', () => {
      expect(firstSecondThird(1, 3)).toEqual(1);
      expect(firstSecondThird(1, 4)).toEqual(1);
      expect(firstSecondThird(2, 4)).toEqual(1);
    });
  });

  describe('fillWhitespace tests', function() {
    it('should fill space characters only after the original string with the given character up to the length', () => {
      expect(fillWhitespace('    ', '    ', '-', 4)).toEqual('    ');
      expect(fillWhitespace('    ', '    ', '-', 8)).toEqual('    ----');
    });

    it('should inherit non-space characters from the current string', () => {
      expect(fillWhitespace('    ', '    char', '-', 4)).toEqual('    char');
      expect(fillWhitespace('    ', '    char', '-', 8)).toEqual('    char');
      expect(fillWhitespace('    ', '    c a ', '-', 8)).toEqual('    c-a-');
      expect(fillWhitespace('    ', '     h r', '-', 8)).toEqual('    -h-r');
      expect(fillWhitespace('    ', '        ', '-', 8)).toEqual('    ----');
    });
  });

  describe('handler tests', function() {
    const yargArgs = {
      $0: 'test',
      _: ['test'],
      json: true
    };
    const config = {
      clientId: 'client-id',
      clientSecret: 'client-id',
      hubId: 'hub-id'
    };

    beforeAll(async () => {
      await rimraf('temp/tree/');
    });

    beforeEach(() => {
      jest.resetAllMocks();
    });

    const itemFromTemplate = (template: ItemTemplate): ContentItem => {
      const item = new ContentItem({
        label: template.label,
        status: template.status || Status.ACTIVE,
        id: template.id || template.label,
        folderId: null,
        version: template.version,
        lastPublishedVersion: template.lastPublishedVersion,
        locale: template.locale,
        body: {
          ...template.body,
          _meta: {
            schema: template.typeSchemaUri
          }
        },

        // Not meant to be here, but used later for sorting by repository
        repoId: template.repoId
      });

      return item;
    };

    const createContent = async (basePath: string, template: ItemTemplate): Promise<void> => {
      await promisify(writeFile)(
        join(basePath, template.label + '.json'),
        JSON.stringify(itemFromTemplate(template).toJSON())
      );
    };

    it('should use getDefaultLogPath for LOG_FILENAME with process.platform as default', function() {
      LOG_FILENAME();

      expect(getDefaultLogPath).toHaveBeenCalledWith('item', 'tree', process.platform);
    });

    it('should print nothing if no content is present', async () => {
      await ensureDirectoryExists('temp/tree/empty');

      const argv = {
        ...yargArgs,
        ...config,
        dir: 'temp/tree/empty'
      };

      await handler(argv);

      expect(consoleLogSpy.mock.calls.map(args => args[0]).join('\n')).toMatchSnapshot();
    });

    it('should print a single content item by itself', async () => {
      const basePath = 'temp/tree/single/repo1';
      await ensureDirectoryExists(basePath);

      await promisify(writeFile)(join(basePath, 'dummyFile.txt'), 'ignored');

      await createContent(basePath, {
        label: 'item1',
        id: 'id1',
        repoId: 'repo1',
        body: {},
        typeSchemaUri: 'http://type.com'
      });

      const argv = {
        ...yargArgs,
        ...config,
        dir: 'temp/tree/single'
      };

      await handler(argv);

      expect(consoleLogSpy.mock.calls.map(args => args[0]).join('\n')).toMatchSnapshot();
    });

    it('should print a tree of content items', async () => {
      const basePath = 'temp/tree/multiple/repo1';
      await ensureDirectoryExists(basePath);

      const shared = { typeSchemaUri: 'http://type.com', repoId: 'repo1' };

      await createContent(basePath, { label: 'item1', id: 'id1', body: dependsOn(['id2', 'id3']), ...shared });
      await createContent(basePath, { label: 'item2', id: 'id2', body: dependsOn(['id4', 'id6']), ...shared });
      await createContent(basePath, { label: 'item3', id: 'id3', body: {}, ...shared });
      await createContent(basePath, { label: 'item4', id: 'id4', body: {}, ...shared });
      await createContent(basePath, { label: 'item5', id: 'id5', body: {}, ...shared });
      await createContent(basePath, { label: 'item6', id: 'id6', body: dependsOn(['id5']), ...shared });

      const argv = {
        ...yargArgs,
        ...config,
        dir: 'temp/tree/multiple'
      };

      await handler(argv);

      expect(consoleLogSpy.mock.calls.map(args => args[0]).join('\n')).toMatchSnapshot();
    });

    it('should print multiple disjoint trees of content items', async () => {
      const basePath = 'temp/tree/disjoint/repo1';
      await ensureDirectoryExists(basePath);

      const shared = { typeSchemaUri: 'http://type.com', repoId: 'repo1' };

      await createContent(basePath, { label: 'item1', id: 'id1', body: dependsOn(['id2', 'id3']), ...shared });
      await createContent(basePath, { label: 'item2', id: 'id2', body: dependsOn(['id4']), ...shared });
      await createContent(basePath, { label: 'item3', id: 'id3', body: {}, ...shared });
      await createContent(basePath, { label: 'item4', id: 'id4', body: {}, ...shared });

      await createContent(basePath, { label: 'item5', id: 'id5', body: {}, ...shared });
      await createContent(basePath, { label: 'item6', id: 'id6', body: dependsOn(['id5']), ...shared });

      await createContent(basePath, { label: 'item7', id: 'id7', body: {}, ...shared });

      const argv = {
        ...yargArgs,
        ...config,
        dir: 'temp/tree/disjoint'
      };

      await handler(argv);

      expect(consoleLogSpy.mock.calls.map(args => args[0]).join('\n')).toMatchSnapshot();
    });

    it('should detect and print circular dependencies with a double line indicator', async () => {
      const basePath = 'temp/tree/disjoint/repo1';
      await ensureDirectoryExists(basePath);

      const shared = { typeSchemaUri: 'http://type.com', repoId: 'repo1' };

      await createContent(basePath, { label: 'item1', id: 'id1', body: dependsOn(['id2', 'id3']), ...shared });
      await createContent(basePath, { label: 'item2', id: 'id2', body: dependsOn(['id4']), ...shared });
      await createContent(basePath, { label: 'item3', id: 'id3', body: {}, ...shared });
      await createContent(basePath, { label: 'item4', id: 'id4', body: dependsOn(['id1']), ...shared });

      const argv = {
        ...yargArgs,
        ...config,
        dir: 'temp/tree/disjoint'
      };

      await handler(argv);

      expect(consoleLogSpy.mock.calls.map(args => args[0]).join('\n')).toMatchSnapshot();
    });

    it('should detect intertwined circular dependencies with multiple lines with different position', async () => {
      const basePath = 'temp/tree/intertwine/repo1';
      await ensureDirectoryExists(basePath);

      const shared = { typeSchemaUri: 'http://type.com', repoId: 'repo1' };

      await createContent(basePath, { label: 'item1', id: 'id1', body: dependsOn(['id2']), ...shared });
      await createContent(basePath, { label: 'item2', id: 'id2', body: dependsOn(['id3']), ...shared });
      await createContent(basePath, { label: 'item3', id: 'id3', body: dependsOn(['id2', 'id4']), ...shared });
      await createContent(basePath, { label: 'item4', id: 'id4', body: dependsOn(['id1', 'id5']), ...shared });

      await createContent(basePath, { label: 'item5', id: 'id5', body: dependsOn(['id6']), ...shared });
      await createContent(basePath, { label: 'item6', id: 'id6', body: dependsOn(['id5']), ...shared });

      const argv = {
        ...yargArgs,
        ...config,
        dir: 'temp/tree/intertwine'
      };

      await handler(argv);

      expect(consoleLogSpy.mock.calls.map(args => args[0]).join('\n')).toMatchSnapshot();
    });

    it('should print an error when invalid json is found', async () => {
      const basePath = 'temp/tree/invalud/repo1';
      await ensureDirectoryExists(basePath);

      await createContent(basePath, {
        label: 'item1',
        id: 'id1',
        repoId: 'repo1',
        body: {},
        typeSchemaUri: 'http://type.com'
      });
      await promisify(writeFile)(join(basePath, 'badfile.json'), 'not json');

      const argv = {
        ...yargArgs,
        ...config,
        dir: 'temp/tree/invalud'
      };

      await handler(argv);

      expect(consoleLogSpy.mock.calls.map(args => args[0]).join('\n')).toMatchSnapshot();
      expect(consoleErrorSpy.mock.calls.map(args => args[0]).join('\n')).toMatchSnapshot();
    });
  });
});
