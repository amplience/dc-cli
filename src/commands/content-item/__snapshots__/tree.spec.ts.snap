// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`content-item tree command handler tests should detect and print circular dependencies with a double line indicator 1`] = `
"=== LEVEL 2 (1) ===
item6 
└─ item5 

=== LEVEL 1 (3) ===
item3 

item7 

=== CIRCULAR (3) ===
item1 ═════════════════╗
├─ item2               ║
│  └─ item4            ║
│     └─ *** (item1) ══╝
└─ (item3) 

Finished. Circular Dependencies printed: 1"
`;

exports[`content-item tree command handler tests should detect intertwined circular dependencies with multiple lines with different position 1`] = `
"test
test
=== CIRCULAR (6) ===
item5 ══════════════╗
└─ item6            ║
   └─ *** (item5) ══╝

item1 ══════════════════════╗
└─ item2 ═════════════════╗ ║
   └─ item3               ║ ║
      ├─ *** (item2) ═════╝ ║
      └─ item4              ║
         ├─ *** (item1) ════╝
         └─ (item5) 

Finished. Circular Dependencies printed: 2"
`;

exports[`content-item tree command handler tests should print a single content item by itself 1`] = `
"=== LEVEL 1 (1) ===
item1 

Finished. Circular Dependencies printed: 0"
`;

exports[`content-item tree command handler tests should print a tree of content items 1`] = `
"=== LEVEL 4 (1) ===
item1 
├─ item2 
│  ├─ item4 
│  └─ item6 
│     └─ item5 
└─ item3 

=== LEVEL 3 (1) ===
=== LEVEL 2 (1) ===
=== LEVEL 1 (3) ===
Finished. Circular Dependencies printed: 0"
`;

exports[`content-item tree command handler tests should print an error when invalid json is found 1`] = `
"=== LEVEL 1 (1) ===
item1 

Finished. Circular Dependencies printed: 0"
`;

exports[`content-item tree command handler tests should print an error when invalid json is found 2`] = `"Couldn't read content item at '/Users/rhys/Documents/amplience/dc-cli/temp/tree/invalud/repo1/badfile.json': SyntaxError: Unexpected token o in JSON at position 1"`;

exports[`content-item tree command handler tests should print multiple disjoint trees of content items 1`] = `
"=== LEVEL 3 (1) ===
item1 
├─ item2 
│  └─ item4 
└─ item3 

=== LEVEL 2 (2) ===
item6 
└─ item5 

=== LEVEL 1 (4) ===
item7 

Finished. Circular Dependencies printed: 0"
`;

exports[`content-item tree command handler tests should print nothing if no content is present 1`] = `"Finished. Circular Dependencies printed: 0"`;
